File: 0125-valid-palindrome/0125-valid-palindrome.swift
class Solution {
    func isPalindrome(_ s: String) -> Bool {
        let s = s.lowercased()
        var left = s.startIndex
        var right = s.index(before: s.endIndex)

        while left < right {
            while left < right && !s[left].isNumber && !s[left].isLetter {
                left = s.index(after: left)
            }

            while left < right && !(s[right].isNumber || s[right].isLetter) {
                right = s.index(before: right)
            }

            if left >= right {
                return true
            }

            if s[left] != s[right] {
                return false
            }

            left = s.index(after: left)
            right = s.index(before: right)
        }

        return true
    }
}

File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1

        while left < right:
            while left < right and not (s[left].isalpha() or s[left].isdigit()):
                left += 1
            
            while right > left and not (s[right].isalpha() or s[right].isdigit()):
                right -= 1

            if s[left].lower() == s[right].lower():
                left += 1
                right -= 1
            else:
                return False

        return True